feat(s3): new check s3_bucket_object_public to detect public objects

Synchx00 opened the pull request:
### Context

The current S3 checks in Prowler verify public access at the bucket level (via Bucket Policies or ACLs). However, a bucket can be private while individual objects inside it are public (e.g., via Object ACLs). This PR addresses this gap by adding a check to detect public objects within buckets.

Fix #3463

### Description

I have added a new check: `s3_bucket_object_public`.

**Logic implemented:**
1. Iterates through all S3 buckets.
2. Lists objects in the bucket (up to 100).
3. Randomly samples 3 objects (as suggested in the issue to avoid performance penalties on large buckets).
4. Inspects the ACL of these sampled objects.
5. Returns **FAIL** if an object grants permissions to the `AllUsers` group (URI: `http://acs.amazonaws.com/groups/global/AllUsers`).
6. Returns **PASS** if the sampled objects are private or if the bucket is empty.

**Dependencies:**
- No new external dependencies. Uses standard `boto3` methods (`list_objects_v2`, `get_object_acl`).

### Steps to review

To verify this check manually:

1. **Create a test S3 bucket** and upload two files:
   - One private file.
   - One public file (e.g., `aws s3api put-object-acl --bucket <bucket> --key public.txt --acl public-read`).
2. **Run the specific check** against that bucket:
   ```bash
   python prowler-cli.py aws --check s3_bucket_object_public --resource-arn arn:aws:s3:::<your-bucket-name>

Please add a detailed description of how to review this PR.

### Checklist

- Are there new checks included in this PR? Yes
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [ ] Review if the code is being covered by tests.
- [ ] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [ ] Review if backport is needed.
- [ ] Review if is needed to change the [Readme.md](https://github.com/prowler-cloud/prowler/blob/master/README.md)
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/prowler/CHANGELOG.md), if applicable.

#### UI
- [ ] All issue/task requirements work as expected on the UI
- [ ] Screenshots/Video of the functionality flow (if applicable) - Mobile (X < 640px)
- [ ] Screenshots/Video of the functionality flow (if applicable) - Table (640px > X < 1024px)
- [ ] Screenshots/Video of the functionality flow (if applicable) - Desktop (X > 1024px)
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/ui/CHANGELOG.md), if applicable.

#### API
- [ ] Verify if API specs need to be regenerated.
- [ ] Check if version updates are required (e.g., specs, Poetry, etc.).
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/api/CHANGELOG.md), if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added an S3 control to spot-check bucket objects for public ACL grants (AllUsers/AuthenticatedUsers) and generate per-bucket PASS/FAIL findings (disabled by default).
  * Added configuration to enable the check and tune sampling limits (max objects and sample size).
* **Bug Fixes**
  * Improved reporting when spot-checking is blocked or errors occur by producing clearer MANUAL outcomes and messages.
* **Tests**
  * Added coverage for the new S3 spot-check behavior, including empty buckets, public ACL detection, and error handling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Synchx00 committed:
feat(s3): new check s3_bucket_object_public to detect public objects

jfagoagas commented:
Hi @Synchx00 thanks for this contribution. I apologise for the delay. We'll review it and get back to you as soon as possible.

Thanks for using Prowler 🚀

Synchx00 commented:
Hi @andoniaf, thanks for the detailed feedback!

I completely agree with your concerns regarding the lexicographical order of `ListObjectsV2` and the potential for creating a false sense of security with the current sampling method.

The proposed path forward, making this check configurable (disabled by default), adding sample size controls, and updating the metadata to explicitly state these limitations, sounds like the perfect solution.

I will start working on these changes right away!

Synchx00 committed:
fix(s3): address review feedback on s3_bucket_object_public

Make the check opt-in via three audit_config keys (disabled by default),
lower severity to low, and clarify that this is a spot-check sample.

- prowler/config/config.yaml: add s3_bucket_object_public_enabled
  (False), _max_objects (100), _sample_size (3) with guidance pointing
  to s3_bucket_acl_prohibited for comprehensive coverage.
- Check early-returns when disabled; otherwise reads max_objects and
  sample_size from audit_config. Uses Check_Report_AWS. Compares the
  AllUsers group URI exactly instead of substring-matching. PASS
  message discloses the sample size and recommends disabling ACLs via
  Object Ownership for full assurance.
- metadata.json: Severity -> low; Description, Notes and Recommendation
  disclose the sampling limitation and cross-reference
  s3_bucket_acl_prohibited; RelatedTo links the two checks.
- Unit tests cover: disabled-by-default (no findings), empty bucket
  PASS, private-objects PASS, public-object FAIL, AccessDenied MANUAL,
  other ClientError MANUAL.

danibarranqueroo committed:
Merge remote-tracking branch 'origin/master' into feature/issue-3463-s3-objects-public

# Conflicts:
#	prowler/config/config.yaml

danibarranqueroo committed:
fix(s3): move object sampling to service layer and fix metadata

- Move list_objects_v2/get_object_acl into s3_service (opt-in, threaded)
- Deterministic sampling; also flag AuthenticatedUsers group
- Fix metadata validation (CheckType, Description, RelatedUrl, Url, Category)
- Add changelog entry (#9517) and required IAM permissions

danibarranqueroo committed:
chore: resolve comments and add validators

danibarranqueroo committed:
Merge branch 'master' into feature/issue-3463-s3-objects-public

danibarranqueroo reviewed (approved):
Really good job! Thanks for this and for all the effort ❤️
